src/download.cpp

```diff
@@ -206,6 +206,13 @@ void DownloadFile(const std::string& url, IDownloadSink *sink, int flags)
 
     if ( !filename_set )
     {
+        char ioptionurl[512];
+        DWORD ousize = 512;
+        if (InternetQueryOptionA(conn, INTERNET_OPTION_URL, ioptionurl, &ousize))
+        {
+            urlc.lpszUrlPath = ioptionurl;
```

Comment:
This seems rather hacky. Why not change `GetURLFileName` to take the URL string as its argument and call it as `GetURLFileName(ioptionurl)` (here) or `GetURLFileName(urlc.lpszUrlPath)` (below in else branch) respectively?